server/gateway.go

```diff
@@ -0,0 +1,1707 @@
+				proto = append(proto, accName...)
+				proto = append(proto, ' ')
+				proto = append(proto, sub.subject...)
+				if sub.queue != nil {
+					proto = append(proto, ' ')
+					proto = append(proto, sub.queue...)
+					// For now, just use 1 for the weight
+					proto = append(proto, ' ', '1')
+				}
+				proto = append(proto, CR_LF...)
+			}
+			c.mu.Lock()
+			if c.trace {
+				c.traceOutOp("", proto[:len(proto)-LEN_CR_LF])
+			}
+			c.sendProto(proto, true)
+			c.mu.Unlock()
+		}
+	}
+}
+
+func (s *Server) sendQueueUnsubToGateways(accName string, qsub *subscription) {
+	// If this is a remote queue sub, need to remove from remote queue subs map
+	if qsub.client != nil && qsub.client.typ == ROUTER {
+		s.gateway.Lock()
+		delete(s.gateway.rqs, string(qsub.sid))
+		s.gateway.Unlock()
+	}
+	var (
+		_gws [4]*client
+		gws  = _gws[:0]
+	)
+	s.getInboundGateways(&gws)
+	if len(gws) == 0 {
+		return
+	}
+	var (
+		protoa = [512]byte{}
+		proto  = protoa[:0]
+	)
+	proto = append(proto, rUnsubBytes...)
+	proto = append(proto, accName...)
+	proto = append(proto, ' ')
+	proto = append(proto, qsub.subject...)
+	proto = append(proto, ' ')
+	proto = append(proto, qsub.queue...)
+	proto = append(proto, CR_LF...)
+	for _, c := range gws {
+		c.mu.Lock()
+		if c.trace {
+			c.traceOutOp("", proto[:len(proto)-LEN_CR_LF])
+		}
+		c.sendProto(proto, true)
+		c.mu.Unlock()
+	}
+}
+
+// May send a message to all outbound gateways. It is possible
+// that message is not sent to a given gateway if for instance
+// it is known that this gateway has no interest in account or
+// subject, etc..
+func (c *client) sendMsgToGateways(msg, subject, reply []byte, doQueues bool) {
+	var (
+		gwsa [4]*client
+		gws  = gwsa[:0]
+		s    = c.srv
+	)
+	s.getOutboundGateways(&gws)
+	if len(gws) == 0 {
+		return
+	}
+	var (
+		subj    = string(subject)
+		queuesa = [512]byte{}
+		queues  = queuesa[:0]
+		groups  = map[string]struct{}{}
+		accName = c.acc.Name
+	)
+	if doQueues {
+		// Order the gws by lowest RTT
+		sort.Slice(gws, func(i, j int) bool {
+			return gws[i].getRTTValue() < gws[j].getRTTValue()
+		})
+	}
+	for _, gwc := range gws {
+		// hasInterest is not covering queue subs
+		psubInterest := gwc.hasInterest(accName, subj)
+		if !psubInterest && !doQueues {
+			continue
+		}
+		mh := c.msgb[:msgHeadProtoLen]
+		mh = append(mh, accName...)
+		mh = append(mh, ' ')
+		mh = append(mh, subject...)
+		mh = append(mh, ' ')
+
+		hasQueues := false
+		if doQueues {
+			queues = queuesa[:0]
+			sli, _ := gwc.gw.qsubsInterest.Load(accName)
+			if sli != nil {
+				sl := sli.(*Sublist)
+				r := sl.Match(subj)
+				if len(r.qsubs) > 0 {
+					for i := 0; i < len(r.qsubs); i++ {
+						qsubs := r.qsubs[i]
+						if len(qsubs) > 0 {
+							queue := qsubs[0].queue
+							if _, exists := groups[string(queue)]; !exists {
+								groups[string(queue)] = struct{}{}
+								queues = append(queues, queue...)
+								queues = append(queues, ' ')
+								hasQueues = true
```

Comment:
Again long indentation usually points to complicated data structures. Sometimes necessary, but good to check.